File: fourier-methods/cpp-implementation/car-madan/tests/polynomial_mult.cpp
/**
 * @file polynomial_mult.cpp
 *
 * @brief Test function uses to evaluate the implementation of the FFT and IFFT
 * using the concept of polynomial multiplication.
 *
 * @author Harsh Parikh
 * @date 8th July 2025
 */
#include <iostream>
#include <vector>
#include <complex>
#include <cmath>

#include "../src/fourier_transform.hpp"
#include "../src/helper_functions.hpp"

/**
 * @brief Tests the FFT-based polynomial multiplication implemented in the `src` folder.
 *
 * This function multiplies two polynomials, represented as vectors of coefficients,
 * using the Fast Fourier Transform (FFT) technique. It serves as a verification tool
 * for the correctness and efficiency of the FFT implementation in the codebase.
 *
 * @param poly_A A vector representing the coefficients of the first polynomial,
 *               in ascending order of degree (i.e., poly_A[i] corresponds to x^i).
 * @param poly_B A vector representing the coefficients of the second polynomial,
 *               also in ascending order of degree.
 *
 * @return A vector containing the coefficients of the resulting polynomial product,
 *         computed via FFT-based convolution.
 */
std::vector<std::complex<double>> multiply_polynomials(const std::vector<double> &poly_A,
                                                       const std::vector<double> &poly_B)
{
    // Calculating the total size
    int result_size = poly_A.size() + poly_B.size() - 1; // -1 accounts for indexing
                                                         // starting at 0.

    // Calculating the next power of two (the size of fft must be a power of 2)
    int N = next_power_of_two(result_size);

    // Converting the vectors into complex datatype for function compatibility
    std::vector<std::complex<double>> A_complex(N, 0), B_complex(N, 0);

    // Populating the vector values into their corresponding complex equivalent.

    // Given that the size of complex vector is `N`, some elements would be stored
    // as `0` and the code would function as intended.
    for (int i = 0; i < poly_A.size(); ++i)
    {
        A_complex[i] = std::complex<double>(poly_A[i], 0.0);
    }

    for (int j = 0; j < poly_B.size(); ++j)
    {
        B_complex[j] = std::complex<double>(poly_B[j], 0.0);
    }

    // Converting the vectors into forward FFT representation
    auto A_fft = discrete_fourier_transform(A_complex);
    auto B_fft = discrete_fourier_transform(B_complex);

    // Multiplying the vectors point-wise
    std::vector<std::complex<double>> C_fft(N);

    for (int i = 0; i < N; ++i)
    {
        C_fft[i] = A_fft[i] * B_fft[i];
    }

    // Calculating the inverse of the product of two vectors
    auto C_time = inverse_fourier_transform(C_fft);

    return C_time;
}

int main()
{
    // Initialising dummy vectors
    std::vector<double> poly_A = {1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1};
    std::vector<double> poly_B = {1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1};

    int result_size = poly_A.size() + poly_B.size() - 1;

    // Inputing the dummy vectors into the function
    auto result = multiply_polynomials(poly_A, poly_B);

    // Printing the result (real-part)
    std::cout << "Resultant Coefficients: \n";
    for (int i = 0; i < result_size; ++i) // The iteration will only be carried
                                          // till the result size as the rest
                                          // of the elements would be 0.
    {
        std::cout << round(result[i].real()) << " ";
    }
    return 0;
}

File: fourier-methods/python-implementation/polynomial_multiplication/polynomial_mult.py
"""
A script for polynomial multiplication using numpy's fft library.

Author: Harsh Parikh
Date: 30th July 2025
"""

import numpy as np


def naive_multiplication(poly_A, poly_B):
    """
    The brute force method of polynomial multiplication.

    Input Parameters
    ----------------
     poly_A : np.ndarray
        Coefficients of the first polynomial, in increasing order of powers (a_0, a_1, ..., a_n).
    poly_B : np.ndarray
        Coefficients of the second polynomial, in increasing order of powers (b_0, b_1, ..., b_m).

    Returns
    -------
    result: np.ndarray
        Coefficients of the product polynomial C(x), of length n + m - 1.
    """
    # Calculating the resultant size
    result_size = len(poly_A) + len(poly_B) - 1

    # Initialising the resultant vector
    result = np.zeros((result_size))

    # Multiplying the polynomials using the naive multiplication method
    for i in range(len(poly_A)):
        for j in range(len(poly_B)):
            result[i + j] += poly_A[i] * poly_B[i]

    return result


def next_power_of_two(n):
    """
    Calculates the next power of two after the number `n`.
    """
    result = 1
    while result < n:
        result <<= 1  # This LOC shifts the bit to the left until the power of two
        # is greater than the input number.
    return result


def multiply_polynomials(poly_A, poly_B):
    """
    Multiply two real-coefficient polynomials using the Fast Fourier Transform (FFT).

    This function takes two 1D NumPy arrays representing the coefficients of polynomials
    A(x) and B(x), and computes their product C(x) = A(x) * B(x) efficiently using
    the convolution theorem and FFT.

    Input Parameters
    ----------
    poly_A : np.ndarray
        Coefficients of the first polynomial, in increasing order of powers (a_0, a_1, ..., a_n).
    poly_B : np.ndarray
        Coefficients of the second polynomial, in increasing order of powers (b_0, b_1, ..., b_m).

    Returns
    -------
    result : np.ndarray
        Real-valued coefficients of the product polynomial C(x), of length n + m - 1.
    """
    # Calculating the total size of the output vector
    result_size = (
        len(poly_A) + len(poly_B) - 1
    )  # -1 accounts for indexing starting at 0.

    # Calculating the next power of two (the size of fft vector must be a power of 2)
    N = next_power_of_two(result_size)

    # Converting the vectors into complex datatype
    A_complex = np.zeros((N), dtype=np.complex64)
    B_complex = np.zeros((N), dtype=np.complex64)

    # Assigning the original vectors to their complex counterparts
    A_complex[0 : len(poly_A)] = poly_A
    B_complex[0 : len(poly_B)] = poly_B

    # Converting the vectors into their forward FFT representation
    A_fft = np.fft.fft(A_complex)
    B_fft = np.fft.fft(B_complex)

    # Multiplying the vectors point-wise
    C_fft = A_fft * B_fft

    # Calculating the inverse of the product of two vectors
    result = np.fft.ifft(C_fft).real

    # The vector is only returned till result size to avoid `0`s
    return result[:result_size]


if __name__ == "__main__":
    vector_A = np.array([1, 1, 1])
    vector_B = np.array([1, 1, 1])

    result = multiply_polynomials(vector_A, vector_B)
    print(result)
